improver/utilities/cube_checker.py

```diff
@@ -203,3 +203,41 @@ def spatial_coords_match(cubes: Union[List, CubeList]) -> bool:
             and match
         )
     return match
+
+
+def assert_time_coords_valid(inputs: List[Cube], time_bounds: bool):
+    """
+    Raises appropriate ValueError if
+
+    - Any input cube has or is missing time bounds (depending on time_bounds)
+    - Input cube times and forecast_reference_times do not match
+
+    Args:
+        inputs:
+            List of Cubes where times should match
+        time_bounds:
+            When True, time-bounds are checked for and compared on the input cubes.
+            When False, the absence of time-bounds is checked for.
+
+    Raises:
+        ValueError: If any of the stated checks fail.
+    """
+    if len(inputs) <= 1:
+        raise ValueError(f"Need at least 2 cubes to check. Found {len(inputs)}")
+    cubes_not_matching_time_bounds = [
+        c.name() for c in inputs if c.coord("time").has_bounds() != time_bounds
+    ]
+    if cubes_not_matching_time_bounds:
+        str_bool = "" if time_bounds else "not "
+        msg = f"{' and '.join(cubes_not_matching_time_bounds)} must {str_bool}have time bounds"
+        raise ValueError(msg)
+    for time_coord_name in ["time", "forecast_reference_time"]:
+        time_coords = [c.coord(time_coord_name) for c in inputs]
+        if not all([tc == time_coords[0] for tc in time_coords[1:]]):
+            raise ValueError(
+                f"{time_coord_name} coordinates do not match."
+                "\n  "
+                "\n  ".join(
+                    [str(c.name()) + ": " + str(c.coord("time")) for c in inputs]
+                )
```

Comment:
I'm not sure that this is functioning as expected, and the full output for this error message isn't captured in any of the tests. The result of this currently is shown below:

> ValueError: air_temperature: DimCoord([2017-11-10 03:00:00], standard_name='time', calendar='gregorian')time coordinates do not match.
>                 
> air_pressure: DimCoord([2017-11-10 04:00:00], standard_name='time', calendar='gregorian')time coordinates do not match.
>                 
> relative_humidity: DimCoord([2017-11-10 04:00:00], standard_name='time',  calendar='gregorian')
>

Do you actually mean to use "time coordinates do not match" as part of the join statement? Or do you mean to state "time coordinates do not match" and then list of coordinates e.g. as below?

> ValueError: time coordinates do not match.
> air_temperature: DimCoord([2017-11-10 03:00:00], standard_name='time', calendar='gregorian')
>                 
> air_pressure: DimCoord([2017-11-10 04:00:00], standard_name='time', calendar='gregorian')
>                 
> relative_humidity: DimCoord([2017-11-10 04:00:00], standard_name='time',  calendar='gregorian')

Reply:
Well spotted! I've added `"^"` to the start of the expected error messages to tell Regex that these should be the start of the matched strings. This traps the mistake you've highlighted here. The mistake is fixed by adding `"+"` between the two new-line characters.